expression/src/dependencies.rs

```diff
@@ -0,0 +1,113 @@
+use crate::error::{self, ExpressionExecutionError};
+use cargo::{
+    core::{
+        compiler::{BuildConfig, MessageFormat},
+        Shell, Workspace,
+    },
+    ops::CompileOptions,
+    util::command_prelude::CompileMode,
+};
+use snafu::{OptionExt, ResultExt, Whatever};
+use std::path::{Path, PathBuf};
+
+pub type Result<T, E = ExpressionExecutionError> = std::result::Result<T, E>;
+
+const DEPS_CARGO_TOML: &[u8] = std::include_bytes!("../deps-workspace/Cargo.toml");
+const DEPS_CARGO_LOCK: &[u8] = std::include_bytes!("../deps-workspace/Cargo.lock");
+const DEPS_LIB_RS: &[u8] = std::include_bytes!("../deps-workspace/lib.rs");
+
+/// A pre-built workspace for linking dependencies.
+///
+/// If this is dropped, the workspace will be deleted.
+///
+pub struct ExpressionDependencies {
+    // we need to hold this to keep the tempdir alive
+    _cargo_workspace: tempfile::TempDir,
+    dependencies: PathBuf,
+}
+
+impl ExpressionDependencies {
+    pub fn new() -> Result<Self> {
+        let cargo_workspace = tempfile::tempdir().context(error::TempDir)?;
+
+        Self::copy_deps_workspace(cargo_workspace.path()).context(error::DepsWorkspace)?;
+
+        let dependencies = Self::cargo_build(cargo_workspace.path()).map_err(|e| {
+            ExpressionExecutionError::DepsBuild {
+                debug: format!("{e:?}"),
+            }
+        })?;
+
+        Ok(Self {
+            _cargo_workspace: cargo_workspace,
+            dependencies,
+        })
+    }
+
+    pub fn linker_path(&self) -> &Path {
+        self.dependencies.as_path()
+    }
+
+    fn copy_deps_workspace(cargo_workspace: &Path) -> Result<(), std::io::Error> {
+        std::fs::write(cargo_workspace.join("Cargo.toml"), DEPS_CARGO_TOML)?;
+        std::fs::write(cargo_workspace.join("Cargo.lock"), DEPS_CARGO_LOCK)?;
+        std::fs::write(cargo_workspace.join("lib.rs"), DEPS_LIB_RS)?;
+        Ok(())
+    }
+
+    /// Builds the dependencies workspace.
+    /// We will use the libraries in the `target/release/deps/` folder.
+    ///
+    /// We return [`Whatever`] since [`cargo::util::errors::CargoResult`] has a lifetime.
```

Comment:
out of interest: whats Whatever?

Reply:
Snafu offers an error type which is just some string with an optional source error.
I found this convenient for the CargoResult :shrug:.